### PR TITLE
make include/exclude easier to use with empty but not null arguments

### PR DESCRIPTION
**Pull request by SylvainJuge**
Fixes a few other occurrences of the same bug fixed in https://github.com/open-telemetry/opentelemetry-java/pull/8177.

When using declarative configuration, we get empty collections when an attribute is not defined, and not `null`, when the `include` statement is omitted this will include none and thus exclude all.

I have found a few other usages of `IncludeExcludePredicate` that have the same bug as what was fixed in #8177, so I think it is better to provide a more generic solution, for example visible in this intermediate commit : https://github.com/open-telemetry/opentelemetry-java/pull/8185/changes/c0e641d60b270979447f3129cfbabaad9f324b14 (this has now been replaced by the generic solution).

This change makes `IncludeExcludePattern` require that at least one of include and exclude arguments is non-null or empty, in other words we need to have at least one exclusion or one inclusion.

### CODE REVIEW

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/ViewBuilder.java [hunk 1]
```diff
@@ -74,6 +74,11 @@ public ViewBuilder setAggregation(Aggregation aggregation) {
    */
   public ViewBuilder setAttributeFilter(Set<String> keysToRetain) {
```
Comment (SylvainJuge) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/ViewBuilder.java:76]: [for reviewer] this is the only place where we use `IncludeExcludePredicate` to potentially remove all items, dealing with this as an exception allows to make most common use-case simpler to call without having to convert empty collection to `null`.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java [hunk 2]
```diff
@@ -57,7 +57,12 @@ private IncludeExcludePredicate(
   /**
    * Create a (case-sensitive) exact matching include exclude predicate.
    *
-   * @throws IllegalArgumentException if {@code included} AND {@code excluded} are null.
+   * <p>When {@code included} is empty or {@literal null}, all values are included by default.
+   *
+   * <p>When {@code excluded} is empty or {@literal null}, no value is excluded by default.
+   *
+   * @throws IllegalArgumentException if both {@code included} AND {@code excluded} are null or
```
Comment (robsunday) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java]: It looks to me like some doc/code improvement is needed here. According to JavaDoc:

- `included` null or empty means that all values are included
- `excluded` null or empty means that no value is excluded, which means all values allowed by `included` are included

Do I get it correctly? If so then why to throw an exception in such a case, that actually seems to be a valid case of including all values?

Comment (SylvainJuge) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java]: I think you are right, throwing an exception is something we can avoid:
- for the most common use-case, the values for `included` and `excluded` come from declarative configuration, so if both `included` and `excluded` are omitted (or explicitly empty `[]`) by the user nothing is explicitly included or excluded and we should include by default. Doing this allows to avoid having implement and test the inclusion by default behavior in lots of places.
- when `IncludeExcludePredicate` is used to potentially "filter all" like in the metric views, then we need to deal with this corner case and only use `IncludeExcludePredicate` when there is at least one exclusion or one inclusion. This use-case is less common and thus I think it's fine not to optimize for it, and such behavior should be covered by tests.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java [hunk 3]
```diff
@@ -38,26 +38,28 @@ private IncludeExcludePredicate(
...
     }
   }
 
   /**
    * Create a (case-sensitive) exact matching include exclude predicate.
    *
-   * @throws IllegalArgumentException if {@code included} AND {@code excluded} are null.
+   * <p>When {@code included} is empty or {@literal null}, all values are included by default.
+   *
+   * <p>When {@code excluded} is empty or {@literal null}, no value is excluded by default.
```
Comment (jack-berg) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java]: #nit
```suggestion
   * <p>When {@code included} is empty or {@literal null}, all values are included.
   *
   * <p>When {@code excluded} is empty or {@literal null}, no value are excluded.
```

Applies below as well.

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ViewFactory.java [hunk 4]
```diff
@@ -35,9 +35,7 @@ public View create(ViewStreamModel model, DeclarativeConfigContext context) {
     if (attributeKeys != null) {
       List<String> included = attributeKeys.getIncluded();
       List<String> excluded = attributeKeys.getExcluded();
-      if (included != null || excluded != null) {
-        builder.setAttributeFilter(IncludeExcludePredicate.createExactMatching(included, excluded));
-      }
+      builder.setAttributeFilter(IncludeExcludePredicate.createExactMatching(included, excluded));
```
Comment (jack-berg) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ViewFactory.java:38]: 👍 ugh this was bad

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java [hunk 5]
```diff
@@ -38,26 +38,28 @@ private IncludeExcludePredicate(
       @Nullable Collection<String> excluded,
       boolean globMatchingEnabled) {
     this.globMatchingEnabled = globMatchingEnabled;
-    this.included = included == null ? null : new LinkedHashSet<>(included);
-    this.excluded = excluded == null ? null : new LinkedHashSet<>(excluded);
+    this.included = copyIfNotEmpty(included);
+    this.excluded = copyIfNotEmpty(excluded);
```
Comment (jack-berg) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java:42]: Here you're equating empty as the same as null. So empty = include all.

In declarative config, this is impossible because [IncludeExclude](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/schema-docs.md#includeexclude-) specified `minItems: 1`. I.e. its either null or present with at least one item. In practice, we don't validate this in the declarative config factory classes. Should probably extract out a dedicated IncludeExcludeFactory to convert IncludeExcludeModel to Predicate<String> so we can consistently enforce size > 0.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/ViewBuilder.java [hunk 6]
```diff
@@ -74,6 +74,11 @@ public ViewBuilder setAggregation(Aggregation aggregation) {
...
     Objects.requireNonNull(keysToRetain, "keysToRetain");
+    if (keysToRetain.isEmpty()) {
+      // include/exclude predicate requires to include or exclude at least one, we have to skip it
+      // when we don't want to include any key.
+      return setAttributeFilter(s -> false);
```
Comment (jack-berg) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/ViewBuilder.java]: We need to translate this to a predicate that has a toString implementation.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java (hunk 5 above)
Comment (SylvainJuge) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java:42]: While the `minItems: 1` in declarative config should prevent this, I think it's simpler to cover a more general use-case as it's likely to be reused elsewhere.

What are the downsides of empty and null being different here ?

In `ResourceFactoryTest` we have the following comment that implies that "empty" should mean "include all": https://github.com/open-telemetry/opentelemetry-java/blob/923098b31cc5e6ab29c357804e9d544e26072a34/sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ResourceFactoryTest.java#L145

In `ResourceFactory` we have another instance that could be simplified if we keep the implementation from this PR.

https://github.com/open-telemetry/opentelemetry-java/blob/5035876ac301ff56ad2af352bbea41422ee48517/sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ResourceFactory.java#L96

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/ViewBuilder.java (hunk 6 above)
Comment (SylvainJuge) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/ViewBuilder.java]: I would suggest to return `IncludeExcludePredicate.createPatternMatching(null, Collections.singletonList("*"))`, this would effectively remove all attributes and provide a `toString` implementation.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java (hunk 5 above)
Comment (jack-berg) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/IncludeExcludePredicate.java:42]: > What are the downsides of empty and null being different here ?

I think this behavior is fine. Maybe not entirely intuitive to everyone, but I don't think there is a behavior which is perfectly intuitive. What I'm saying is that the declarative config implementation needs to enforce the constraints of the schema, so that at least in the case of declarative config, an empty include / exclude list never happens in practice, and user's never have the chance to encounter potentially unintuitive behavior.

**Review by jack-berg — APPROVED** [review]
